core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactory.js

```diff
@@ -0,0 +1,239 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Factory for creating instances of NumberWithUnits and Units
+ * domain objects.
+ */
+
+oppia.constant('NUMBER_WITH_UNITS_PARSING_ERRORS', {
+  INVALID_VALUE:
+    'Please ensure that value is either a fraction or a number',
+  INVALID_CURRENCY:
+    'Please enter a valid currency (e.g., $ 5)',
+  INVALID_UNIT_CHARS:
+    'Please ensure that unit only contains numbers, alphabets, (, ), *, ^, /, -'
+});
+
+oppia.factory('NumberWithUnitsObjectFactory', [
+  'UnitsObjectFactory', 'FractionObjectFactory',
+  'NUMBER_WITH_UNITS_PARSING_ERRORS', function(
+      UnitsObjectFactory, FractionObjectFactory,
+      NUMBER_WITH_UNITS_PARSING_ERRORS) {
+    var NumberWithUnits = function(type, real, fraction, units) {
+      this.type = type;
+      this.real = real;
+      this.fraction = fraction;
+      this.units = units;
+    };
+
+    NumberWithUnits.prototype.toString = function() {
+      var numberWithUnitsString = '';
+      if (this.type === 'real') {
+        if (this.units.trim() === '$' || this.units.trim() === 'Rs') {
+          numberWithUnitsString += this.units + ' ' + this.real;
+        } else {
+          numberWithUnitsString += this.real + ' ' + this.units;
+        }
+      } else {
+        if (this.units.trim() === '$' || this.units.trim() === 'Rs') {
+          numberWithUnitsString += this.units + ' ' + this.fraction.toString();
+        } else {
+          numberWithUnitsString += this.fraction.toString() + ' ' + this.units;
+        }
+      }
+      numberWithUnitsString = numberWithUnitsString.trim();
+      return numberWithUnitsString === '' ? '0' : numberWithUnitsString;
+    };
+
+    NumberWithUnits.fromRawInputString = function(rawInput) {
+      rawInput = rawInput.trim();
+      var type = '';
+      var real = '';
+      var Fraction = '';
+      var units = '';
+      var value = 0;
+      var Units = '';
+
+      // Start with digit when there is no currency unit.
+      if (rawInput.match(/^\d/)) {
+        var ind = rawInput.indexOf(' ');
+        if (ind === -1) {
+          // There is value with no units.
+          value = rawInput;
+        } else {
+          value = rawInput.substr(0, ind);
+          units = rawInput.substr(ind + 1).trim();
+        }
+      } else {
+        var ind = rawInput.indexOf(' ');
+        if (ind === -1) {
+          throw new Error(NUMBER_WITH_UNITS_PARSING_ERRORS.INVALID_CURRENCY);
+        }
+        units = rawInput.substr(0, ind) + ' ';
+        var ind2 = rawInput.indexOf(' ', ind + 1);
```

Comment:
A fraction can also be written as '2 / 5' or '3 2 / 5 ', i.e. there can be spaces in between numbers. So modify this part to identify fractions like that.

Reply:
Oh, yes. It's going to be tricky then.

Reply:
Done.